apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -72,6 +81,61 @@ store_message(Msg) ->
 find_subscribers(_Msg) ->
     [node()].
 
+open_session(ClientID) ->
+    ?WHEN_ENABLED(emqx_ds:session_open(ClientID)).
+
+-spec add_subscription(emqx_types:topic(), emqx_ds:session_id()) ->
+    {ok, emqx_ds:iterator_id(), IsNew :: boolean()} | {skipped, disabled}.
+add_subscription(TopicFilterBin, DSSessionID) ->
+    ?WHEN_ENABLED(
+        begin
+            TopicFilter = emqx_topic:words(TopicFilterBin),
+            {ok, IteratorID, StartMS, IsNew} = emqx_ds:session_add_iterator(
+                DSSessionID, TopicFilter
+            ),
+            Ctx = #{
+                iterator_id => IteratorID,
+                start_time => StartMS,
+                is_new => IsNew
+            },
+            ?tp(persistent_session_ds_iterator_added, Ctx),
+            ?tp_span(
+                persistent_session_ds_open_iterators,
+                Ctx,
+                ok = open_iterator_on_all_shards(TopicFilter, StartMS, IteratorID)
+            ),
+            {ok, IteratorID, IsNew}
+        end
+    ).
+
+-spec open_iterator_on_all_shards(emqx_topic:words(), emqx_ds:time(), emqx_ds:iterator_id()) -> ok.
+open_iterator_on_all_shards(TopicFilter, StartMS, IteratorID) ->
+    ?tp(persistent_session_ds_will_open_iterators, #{
+        iterator_id => IteratorID,
+        start_time => StartMS
+    }),
+    %% Note: currently, shards map 1:1 to nodes, but this will change in the future.
+    Nodes = emqx:running_nodes(),
+    Results = emqx_persistent_session_ds_proto_v1:open_iterator(
+        Nodes, TopicFilter, StartMS, IteratorID
+    ),
+    %% TODO: handle errors
+    true = lists:all(fun(Res) -> Res =:= {ok, ok} end, Results),
+    ok.
+
+-spec do_open_iterator(emqx_topic:words(), emqx_ds:time(), emqx_ds:iterator_id()) -> ok.
+do_open_iterator(TopicFilter, StartMS, IteratorID) ->
+    Replay = {TopicFilter, StartMS},
+    case emqx_ds_storage_layer:is_iterator_present(?DS_SHARD, IteratorID) of
+        true ->
+            {ok, _It} = emqx_ds_storage_layer:restore_iterator(?DS_SHARD, IteratorID),
+            ok;
+        false ->
+            {ok, It} = emqx_ds_storage_layer:make_iterator(?DS_SHARD, Replay),
+            ok = emqx_ds_storage_layer:preserve_iterator(It, IteratorID),
+            ok
+    end.
```

Comment:
Nit. I think this is somewhat unnecessary as it calls `iterator_get_state/2` twice for no reason. We could as well have this instead:
```
    case emqx_ds_storage_layer:restore_iterator(?DS_SHARD, IteratorID) of
        {ok, _It} ->
            ok;
        {error, not_found} ->
            {ok, It} = emqx_ds_storage_layer:make_iterator(?DS_SHARD, Replay),
            ok = emqx_ds_storage_layer:preserve_iterator(It, IteratorID),
            ok
    end.
```

Which also renders `is_iterator_present/2` obsolete. Still, this is kinda messy on that level, maybe hide that behind something like `emqx_ds_storage_layer:ensure_iterator/3` which does the same.

Tangential question: do we need to additionally assert that restored iterator was made with same `Replay`, and error out if there's an inconsistency?

Reply:
Good point, `is_iterator_present` apparently became obsolete after some iterations on the solution.  I remember I needed to introduce it initially, but can't remember why exactly.  :see_no_evil: